src/bp-core/compatibility/bp-wp-offload-media-helpers.php

```diff
@@ -0,0 +1,325 @@
+
+		add_action( 'bb_before_document_upload_handler', array( $this, 'bb_offload_media_set_private' ) );
+		add_action( 'bb_before_media_upload_handler', array( $this, 'bb_offload_media_set_private' ) );
+		add_action( 'bb_before_video_upload_handler', array( $this, 'bb_offload_media_set_private' ) );
+		add_action( 'bb_before_video_preview_image_by_js', array( $this, 'bb_offload_media_set_private' ) );
+		add_action( 'bb_video_before_preview_generate', array( $this, 'bb_offload_media_set_private' ) );
+		add_action( 'bb_before_bp_video_thumbnail_upload_handler', array( $this, 'bb_offload_media_set_private' ) );
+		add_action( 'bb_document_before_generate_document_previews', array( $this, 'bb_offload_media_set_private' ) );
+
+		add_action( 'bb_after_document_upload_handler', array( $this, 'bb_offload_media_unset_private' ) );
+		add_action( 'bb_after_media_upload_handler', array( $this, 'bb_offload_media_unset_private' ) );
+		add_action( 'bb_after_video_upload_handler', array( $this, 'bb_offload_media_unset_private' ) );
+		add_action( 'bb_after_video_preview_image_by_js', array( $this, 'bb_offload_media_unset_private' ) );
+		add_action( 'bb_video_after_preview_generate', array( $this, 'bb_offload_media_unset_private' ) );
+		add_action( 'bb_after_bp_video_thumbnail_upload_handler', array( $this, 'bb_offload_media_unset_private' ) );
+		add_action( 'bb_document_after_generate_document_previews', array( $this, 'bb_offload_media_unset_private' ) );
+
+		add_filter( 'bp_document_get_preview_url', array( $this, 'bp_document_offload_get_preview_url' ), PHP_INT_MAX, 6 );
+		add_filter( 'bp_media_get_preview_image_url', array( $this, 'bp_media_offload_get_preview_url' ), PHP_INT_MAX, 5 );
+		add_filter( 'bb_video_get_thumb_url', array( $this, 'bp_video_offload_get_thumb_preview_url' ), PHP_INT_MAX, 5 );
+		add_filter( 'bb_video_get_symlink', array( $this, 'bp_video_offload_get_video_url' ), PHP_INT_MAX, 4 );
+
+		add_action( 'bb_try_before_video_background_create_thumbnail', array( $this, 'bb_video_set_wp_offload_download_video_local' ), 99999, 1 );
+		add_action( 'bb_try_after_video_background_create_thumbnail', array( $this, 'bb_video_unset_wp_offload_download_video_local' ), 99999, 1 );
+
+	}
+
+	/**
+	 * Function to set the false to use the default media symlink instead use the offload media URL of media.
+	 *
+	 * @param bool   $can           default true.
+	 * @param int    $id            media/document/video id.
+	 * @param int    $attachment_id attachment id.
+	 * @param string $size          preview size.
+	 *
+	 * @return bool true if the offload media used.
+	 *
+	 * @since BuddyBoss 1.7.0
+	 */
+	public function bb_offload_do_symlink( $can, $id, $attachment_id, $size ) {
+
+		/**
+		 * Global object for Amazon_S3_And_CloudFront|Amazon_S3_And_CloudFront_Pro.
+		 *
+		 * @var Amazon_S3_And_CloudFront|\Amazon_S3_And_CloudFront_Pro $as3cf
+		 */
+		global $as3cf;
+
+		if ( ! $as3cf->is_attachment_served_by_provider( $attachment_id ) ) {
```

Comment:
This is quite an overhead, and is reaching into WP Offload Media to use a function that isn't part of a public API and may go away some time in the future as we make no guarantees about it.

I realise this is here because one of our team advised that this function exists, but I don't think you need to be calling into our plugin like this.

Instead, just return `false` as you know that symlinks can't be used to override the URLs for media when WP Offload Media might be involved.

Personally I would replace the `add_filter` calls to this function in the above `compatibility_init()` function with something like the following as WP Offload Media has to already be active for that function to be called...

```
add_filter( 'bb_media_do_symlink', '__return_false' );
...
```

Reply:
@ianmjones I did changes based on different cases like offload media handling the URL like if they are in the local server then you are providing the local URL if they are offloaded then giving the bucket URL. So, I am trying to do the same thing like if the document is offloaded then will use the offload URL and if it's not serving via the offload media then I am creating the `symlink` or via `rewrite_rules` based on the server support.

So it would be great if we know the media is serving via local URL or via offloaded.

Let me know your thoughts here.

Reply:
> So it would be great if we know the media is serving via local URL or via offloaded.

I don't think this is a good idea. There's all kinds of scenarios whereby an item could be offloaded, removed from bucket, and then offloaded again, with and without a local file still being available after offload. This will cause problems for BuddyBoss when trying to catch up with the changes and trying to determine whether a symlink should exist or not.

BuddyBoss shouldn't care whether a file is offloaded or not. If standard Media Library functions are being used under the hood to get an attachment's URL then WP Offload Media will filter that URL as appropriate to point to the offloaded version. Just as `get_attached_file()` is filtered to ensure a remote path is returned or optionally the file temporarily downloaded when previously removed from server after offload (or never existed on a server in a clustered setup).

As long as BuddyBoss puts in the safety net of turning off symlinks (or advises users to not use symlinks) when WP Offload Media is also being used, and makes sure it has the local files it needs when generating document previews etc, then hopefully there should not be any issues.